runtime/oti/ObjectMonitor.hpp

```diff
@@ -473,6 +482,113 @@ class VM_ObjectMonitor
+		 */
+		UDATA lockReservationHistory = clazz->lockReservationHistory;
+
+		/* If incrementing the Cancel Counter would cause an overflow, first divide Cancel Counter and Reserved Counter by 2. */
+		if ((lockReservationHistory & 0xFFFF) < 0xFFFF) {
+			clazz->lockReservationHistory = lockReservationHistory + 1;
+		} else {
+			clazz->lockReservationHistory = ((lockReservationHistory >> 1) & 0x7FFF7FFF) + 1;
+		}
+	}
+
+	/**
+	 * Determine initial lockword value based on lockReservationHistory stored in the J9Class.
+	 *
+	 * @param javaVM[in] the J9JavaVM
+	 * @param clazz[in] the object's J9Class
+	 *
+	 * @returns The initial lockword value for the object
+	 */
+	static VMINLINE j9objectmonitor_t
+	getInitialLockword(J9JavaVM *javaVM, J9Class *clazz)
+	{
+		j9objectmonitor_t initialLockword = 0;
+
+		if (javaVM->enableGlobalLockReservation) {
+			/* This path is taken when Global Lock Reservation is enabled. */
+			UDATA lockReservationHistory = clazz->lockReservationHistory;
+
+			/*
+			 * Extract reservedCount and cancelCount from lockReservationHistory.
+			 * cancelCount is the lowest 16 bits of lockReservationHistory.
+			 * reservedCount is the second lowest 16 bits of lockReservationHistory.
+			 *
+			 * lockReservationHistory Format:
+			 * | Unused alignment padding (0/32 bits) | Reserved Counter (Unsigned 16 bits) | Cancel Counter (Unsigned 16 bits) |
+			 */
+			U_32 reservedCount = (lockReservationHistory >> 16) & 0xFFFF;
+			U_32 cancelCount = lockReservationHistory & 0xFFFF;
+
+			U_32 reservedAbsoluteThreshold = javaVM->reservedAbsoluteThreshold;
+			U_32 minimumReservedRatio = javaVM->minimumReservedRatio;
+
+			U_32 cancelAbsoluteThreshold = javaVM->cancelAbsoluteThreshold;
+			U_32 minimumLearningRatio = javaVM->minimumLearningRatio;
+
+			/*
+			 * Check reservedCount and cancelCount against different thresholds to determine what to initialize the lockword to.
+			 * First check if the ratio of resevation to cancellations is high enough to set the lockword to the New-AutoReserve state.
+			 * If so, the initial lockword value is OBJECT_HEADER_LOCK_RESERVED.
+			 * Second check if the ratio is high enough to set the lockword to the New-PreLearning state.
+			 * If so, the initial lockword value is OBJECT_HEADER_LOCK_LEARNING.
+			 * If the ratio is too low, the lockword starts in the Flat-Unlocked state and 0 is returned for the lockword value.
+			 *
+			 * Start as New-AutoReserve -> Initial lockword: OBJECT_HEADER_LOCK_RESERVED
+			 * Start as New-PreLearning -> Initial lockword: OBJECT_HEADER_LOCK_LEARNING
+			 * Start as Flat-Unlocked   -> Initial lockword: 0
+			 */
+			if (reservedCount >= reservedAbsoluteThreshold && (reservedCount > (cancelCount * minimumReservedRatio))) {
+				initialLockword = OBJECT_HEADER_LOCK_RESERVED;
+			} else if (cancelCount < cancelAbsoluteThreshold || (reservedCount > (cancelCount * minimumLearningRatio))) {
```

Comment:
please add extra brackets `(cancelCount < cancelAbsoluteThreshold)`

Reply:
I'll add this as well.